### PR TITLE
CoreSMTSolver: sticky polarities

### PR DESCRIPTION
**Pull request by aehyvari**
### With negation, against master in `QF_UF`
 
non-incremental-QF_UF-master-2022-04-04_vs_879f57ac-2022-04-04

### Without negation, against master in `QF_UF`

non-incremental-QF_UF-master-2022-04-04_vs_fa2481fe-2022-04-05

### CODE REVIEW

File: src/smtsolvers/CoreSMTSolver.cc [hunk 1]
```diff
@@ -406,6 +407,12 @@ void CoreSMTSolver::cancelUntil(int level)
 {
     if (decisionLevel() > level)
     {
+        if (trail.size() > longest_trail) {
+            for (auto p : trail) {
+                saved_polar[var(p)] = not sign(p);
```
Comment (blishko) [inline src/smtsolvers/CoreSMTSolver.cc]: Why do you save negation of `sign` here? I thought the idea was to save the discovered polarity and use that, not it's negation.

Comment (msoos) [reply src/smtsolvers/CoreSMTSolver.cc]: Funnily enough, both can be useful(!?), it's weird. Going for its negation can help with UNSAT. However, you are right, the initial intuition is to not negate it... but either can work(!). Strange. Paper:

http://fmv.jku.at/papers/BiereFleury-POS20.pdf

Note `phase ‘I'`, i.e. the inverted phase. However, likely we should try it without inversion, too.

Comment (aehyvari) [reply src/smtsolvers/CoreSMTSolver.cc]: Yes, I think we got a slightly better result yesterday for our one test instance with the negated polarity, and it's the version that was on my disk, so I ran that.  I'll check the "intuitive" version quickly as well.

**Comment by msoos** [comment]
Can you also try https://github.com/aehyvari/OpenSMT2/pull/11 maybe? It's a bit of a more hackish hack :laughing: 

Mate

**Pull request closed without merging**